TiFlash can not start when restart multiple times

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

Use nightly tiflash, and use `tiup cluster restart` multiple times, then tiflash can not start anymore.

<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

restart normally

### 3. What did you see instead (Required)
```
[2022/12/13 21:43:19.347 +08:00] [ERROR] [Exception.cpp:89] ["Code: 0, e.displayText() = DB::Exception: Can not find path for DMFile, file_id=94893, e.what() = DB::Exception, Stack trace:\n\n\n       0x1708073\tStackTrace::StackTrace() [tiflash+24150131]\n       0x1705e76\tDB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) [tiflash+24141430]\n       0x68be70c\tDB::StableDiskDelegator::getDTFilePath(unsigned long, bool) const [tiflash+109831948]\n       0x6a0df89\tDB::DM::StableValueSpace::restore(DB::DM::DMContext&, unsigned long) [tiflash+111206281]\n       0x69a6849\tDB::DM::Segment::restoreSegment(std::__1::shared_ptr<DB::Logger> const&, DB::DM::DMContext&, unsigned long) [tiflash+110782537]\n       0x6949090\tDB::DM::DeltaMergeStore::DeltaMergeStore(DB::Context&, bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, long, bool, std::__1::vector<DB::DM::ColumnDefine, std::__1::allocator<DB::DM::ColumnDefine> > const&, DB::DM::ColumnDefine const&, bool, unsigned long, DB::DM::DeltaMergeStore::Settings const&) [tiflash+110399632]\n       0x68d56f1\tDB::StorageDeltaMerge::getAndMaybeInitStore() [tiflash+109926129]\n       0x68d96c8\tDB::StorageDeltaMerge::flushCache(DB::Context const&, DB::DM::RowKeyRange const&, bool) [tiflash+109942472]\n       0x6b3b35a\tDB::KVStore::tryFlushRegionCacheInStorage(DB::TMTContext&, DB::Region const&, std::__1::shared_ptr<DB::Logger> const&, bool) [tiflash+112440154]\n       0x6b3e64f\tDB::KVStore::forceFlushRegionDataImpl(DB::Region&, bool, DB::TMTContext&, DB::RegionTaskLock const&, unsigned long, unsigned long) [tiflash+112453199]\n       0x6b3df68\tDB::KVStore::canFlushRegionDataImpl(std::__1::shared_ptr<DB::Region> const&, unsigned char, bool, DB::TMTContext&, DB::RegionTaskLock const&, unsigned long, unsigned long) [tiflash+112451432]\n       0x6b3e3c1\tDB::KVStore::tryFlushRegionData(unsigned long, bool, bool, DB::TMTContext&, unsigned long, unsigned long) [tiflash+112452545]\n       0x6b5b807\tTryFlushData [tiflash+112572423]\n  0x7f4e55f62a32\t_$LT$engine_store_ffi..observer..TiFlashObserver$u20$as$u20$raftstore..coprocessor..AdminObserver$GT$::pre_exec_admin::h39bffcdc6224fe8d [libtiflash_proxy.so+17558066]\n  0x7f4e56e726dd\traftstore::store::fsm::apply::ApplyDelegate$LT$EK$GT$::apply_raft_cmd::h455e45aa599d2a64 [libtiflash_proxy.so+33351389]\n  0x7f4e56e8b955\traftstore::store::fsm::apply::ApplyDelegate$LT$EK$GT$::process_raft_cmd::hb3499b22373e583d [libtiflash_proxy.so+33454421]\n  0x7f4e56e90e24\traftstore::store::fsm::apply::ApplyDelegate$LT$EK$GT$::handle_raft_committed_entries::h0e79a570c1792a09 [libtiflash_proxy.so+33476132]\n  0x7f4e56e6640c\traftstore::store::fsm::apply::ApplyFsm$LT$EK$GT$::handle_apply::h7ef360f580b34c40 [libtiflash_proxy.so+33301516]\n  0x7f4e56e69a38\traftstore::store::fsm::apply::ApplyFsm$LT$EK$GT$::handle_tasks::hbdf512f0cd58e33c [libtiflash_proxy.so+33315384]\n  0x7f4e56508494\t_$LT$raftstore..store..fsm..apply..ApplyPoller$LT$EK$GT$$u20$as$u20$batch_system..batch..PollHandler$LT$raftstore..store..fsm..apply..ApplyFsm$LT$EK$GT$$C$raftstore..store..fsm..apply..ControlFsm$GT$$GT$::handle_normal::h1a10c2992601e20d [libtiflash_proxy.so+23479444]\n  0x7f4e564a38c7\tbatch_system::batch::Poller$LT$N$C$C$C$Handler$GT$::poll::h1e6b33b2fb50cbab [libtiflash_proxy.so+23066823]\n  0x7f4e56559d82\tstd::sys_common::backtrace::__rust_begin_short_backtrace::h176c5187879ecf12 [libtiflash_proxy.so+23813506]\n  0x7f4e565999be\tcore::ops::function::FnOnce::call_once$u7b$$u7b$vtable.shim$u7d$$u7d$::h4eea01b9a3c057b7 [libtiflash_proxy.so+24074686]\n  0x7f4e5753f735\tstd::sys::unix::thread::Thread::new::thread_start::hd2791a9cabec1fda [libtiflash_proxy.so+40482613]\n                \t/rustc/96ddd32c4bfb1d78f0cd03eb068b1710a8cebeef/library/std/src/sys/unix/thread.rs:108\n  0x7f4e54a87ea5\tstart_thread [libpthread.so.0+32421]\n  0x7f4e54093b0d\tclone [libc.so.6+1043213]"] [source="DB::DM::DeltaMergeStore::DeltaMergeStore(DB::Context &, bool, const DB::String &, const DB::String &, DB::TableID, bool, const DB::DM::ColumnDefines &, const DB::DM::ColumnDefine &, bool, size_t, const DB::DM::DeltaMergeStore::Settings &)"] [thread_id=50]
[2022/12/13 21:43:19.394 +08:00] [INFO] [DeltaMergeStore.cpp:209] ["Restore DeltaMerge Store start"] [source="table_id=93"] [thread_id=51]
[2022/12/13 21:43:19.394 +08:00] [DEBUG] [DeltaMergeStore.cpp:1483] ["Loading dt files"] [source="table_id=93"] [thread_id=51]
[2022/12/13 21:43:19.394 +08:00] [ERROR] [Exception.cpp:89] ["Code: 0, e.displayText() = DB::Exception: Can not find path for DMFile, file_id=94893, e.what() = DB::Exception, Stack trace:\n\n\n       0x1708073\tStackTrace::StackTrace() [tiflash+24150131]\n       0x1705e76\tDB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) [tiflash+24141430]\n       0x68be70c\tDB::StableDiskDelegator::getDTFilePath(unsigned long, bool) const [tiflash+109831948]\n       0x6a0df89\tDB::DM::StableValueSpace::restore(DB::DM::DMContext&, unsigned long) [tiflash+111206281]\n       0x69a6849\tDB::DM::Segment::restoreSegment(std::__1::shared_ptr<DB::Logger> const&, DB::DM::DMContext&, unsigned long) [tiflash+110782537]\n       0x6949090\tDB::DM::DeltaMergeStore::DeltaMergeStore(DB::Context&, bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, long, bool, std::__1::vector<DB::DM::ColumnDefine, std::__1::allocator<DB::DM::ColumnDefine> > const&, DB::DM::ColumnDefine const&, bool, unsigned long, DB::DM::DeltaMergeStore::Settings const&) [tiflash+110399632]\n       0x68d56f1\tDB::StorageDeltaMerge::getAndMaybeInitStore() [tiflash+109926129]\n       0x68d96c8\tDB::StorageDeltaMerge::flushCache(DB::Context const&, DB::DM::RowKeyRange const&, bool) [tiflash+109942472]\n       0x6b3b35a\tDB::KVStore::tryFlushRegionCacheInStorage(DB::TMTContext&, DB::Region const&, std::__1::shared_ptr<DB::Logger> const&, bool) [tiflash+112440154]\n       0x6b3e64f\tDB::KVStore::forceFlushRegionDataImpl(DB::Region&, bool, DB::TMTContext&, DB::RegionTaskLock const&, unsigned long, unsigned long) [tiflash+112453199]\n       0x6b3df68\tDB::KVStore::canFlushRegionDataImpl(std::__1::shared_ptr<DB::Region> const&, unsigned char, bool, DB::TMTContext&, DB::RegionTaskLock const&, unsigned long, unsigned long) [tiflash+112451432]\n       0x6b3e3c1\tDB::KVStore::tryFlushRegionData(unsigned long, bool, bool, DB::TMTContext&, unsigned long, unsigned long) [tiflash+112452545]\n       0x6b5b807\tTryFlushData [tiflash+112572423]\n  0x7f4e55f62a32\t_$LT$engine_store_ffi..observer..TiFlashObserver$u20$as$u20$raftstore..coprocessor..AdminObserver$GT$::pre_exec_admin::h39bffcdc6224fe8d [libtiflash_proxy.so+17558066]\n  0x7f4e56e726dd\traftstore::store::fsm::apply::ApplyDelegate$LT$EK$GT$::apply_raft_cmd::h455e45aa599d2a64 [libtiflash_proxy.so+33351389]\n  0x7f4e56e8b955\traftstore::store::fsm::apply::ApplyDelegate$LT$EK$GT$::process_raft_cmd::hb3499b22373e583d [libtiflash_proxy.so+33454421]\n  0x7f4e56e90e24\traftstore::store::fsm::apply::ApplyDelegate$LT$EK$GT$::handle_raft_committed_entries::h0e79a570c1792a09 [libtiflash_proxy.so+33476132]\n  0x7f4e56e6640c\traftstore::store::fsm::apply::ApplyFsm$LT$EK$GT$::handle_apply::h7ef360f580b34c40 [libtiflash_proxy.so+33301516]\n  0x7f4e56e69a38\traftstore::store::fsm::apply::ApplyFsm$LT$EK$GT$::handle_tasks::hbdf512f0cd58e33c [libtiflash_proxy.so+33315384]\n  0x7f4e56508494\t_$LT$raftstore..store..fsm..apply..ApplyPoller$LT$EK$GT$$u20$as$u20$batch_system..batch..PollHandler$LT$raftstore..store..fsm..apply..ApplyFsm$LT$EK$GT$$C$raftstore..store..fsm..apply..ControlFsm$GT$$GT$::handle_normal::h1a10c2992601e20d [libtiflash_proxy.so+23479444]\n  0x7f4e564a38c7\tbatch_system::batch::Poller$LT$N$C$C$C$Handler$GT$::poll::h1e6b33b2fb50cbab [libtiflash_proxy.so+23066823]\n  0x7f4e56559d82\tstd::sys_common::backtrace::__rust_begin_short_backtrace::h176c5187879ecf12 [libtiflash_proxy.so+23813506]\n  0x7f4e565999be\tcore::ops::function::FnOnce::call_once$u7b$$u7b$vtable.shim$u7d$$u7d$::h4eea01b9a3c057b7 [libtiflash_proxy.so+24074686]\n  0x7f4e5753f735\tstd::sys::unix::thread::Thread::new::thread_start::hd2791a9cabec1fda [libtiflash_proxy.so+40482613]\n                \t/rustc/96ddd32c4bfb1d78f0cd03eb068b1710a8cebeef/library/std/src/sys/unix/thread.rs:108\n  0x7f4e54a87ea5\tstart_thread [libpthread.so.0+32421]\n  0x7f4e54093b0d\tclone [libc.so.6+1043213]"] [source="uint8_t DB::TryFlushData(DB::EngineStoreServerWrap *, uint64_t, uint8_t, uint64_t, uint64_t)"] [thread_id=50]
```
![wVGYmx9npu](https://user-images.githubusercontent.com/26538495/207525504-eef227ba-151f-4e9c-a0f6-02b42440c566.jpg)

log is in 172-16-5-81  /data2/hongyunyan/tiflash-41350/log

When I restart that time, there is no workload.

### 4. What is your TiFlash version? (Required)
6.5.0(nightly)

--- a/dbms/src/Interpreters/Context.cpp
+++ b/dbms/src/Interpreters/Context.cpp
@@ -240,6 +240,13 @@ struct ContextShared
             return;
         shutdown_called = true;
 
+        if (global_storage_pool)
+        {
+            // shutdown the gc task of global storage pool before
+            // shutting down the tables.
+            global_storage_pool->shutdown();
+        }
+
         /** At this point, some tables may have threads that block our mutex.
           * To complete them correctly, we will copy the current list of tables,
           *  and ask them all to finish their work.

--- a/dbms/src/Storages/DeltaMerge/DeltaMergeStore.cpp
+++ b/dbms/src/Storages/DeltaMerge/DeltaMergeStore.cpp
@@ -403,10 +403,11 @@ void DeltaMergeStore::shutdown()
         return;
 
     LOG_TRACE(log, "Shutdown DeltaMerge start");
-    // shutdown before unregister to avoid conflict between this thread and background gc thread on the `ExternalPagesCallbacks`
-    // because PageStorage V2 doesn't have any lock protection on the `ExternalPagesCallbacks`.(The order doesn't matter for V3)
+    // Must shutdown storage path pool to make sure the DMFile remove callbacks
+    // won't remove dmfiles unexpectly.
+    path_pool->shutdown();
+    // shutdown storage pool and clean up the local DMFile remove callbacks
     storage_pool->shutdown();
-    storage_pool->dataUnregisterExternalPagesCallbacks(storage_pool->getNamespaceId());
 
     background_pool.removeTask(background_task_handle);
     blockable_background_pool.removeTask(blockable_background_pool_handle);

--- a/dbms/src/Storages/DeltaMerge/DeltaMergeStore.h
+++ b/dbms/src/Storages/DeltaMerge/DeltaMergeStore.h
@@ -52,6 +52,11 @@ using SegmentIdSet = std::unordered_set<UInt64>;
 struct ExternalDTFileInfo;
 struct GCOptions;
 
+namespace tests
+{
+class DeltaMergeStoreTest;
+}
+
 inline static const PageId DELTA_MERGE_FIRST_SEGMENT_ID = 1;
 
 struct SegmentStats
@@ -155,6 +160,7 @@ struct StoreStats
 class DeltaMergeStore : private boost::noncopyable
 {
 public:
+    friend class ::DB::DM::tests::DeltaMergeStoreTest;
     struct Settings
     {
         NotCompress not_compress_columns{};

--- a/dbms/src/Storages/DeltaMerge/DeltaMergeStore_InternalBg.cpp
+++ b/dbms/src/Storages/DeltaMerge/DeltaMergeStore_InternalBg.cpp
@@ -14,12 +14,15 @@
 
 #include <Common/SyncPoint/SyncPoint.h>
 #include <Common/TiFlashMetrics.h>
+#include <Encryption/FileProvider.h>
 #include <Storages/DeltaMerge/DeltaMergeStore.h>
 #include <Storages/DeltaMerge/GCOptions.h>
 #include <Storages/DeltaMerge/Segment.h>
+#include <Storages/PathPool.h>
 #include <Storages/Transaction/TMTContext.h>
 
 #include <magic_enum.hpp>
+#include <memory>
 
 namespace CurrentMetrics
 {
@@ -36,20 +39,33 @@ extern const char pause_until_dt_background_delta_merge[];
 
 namespace DM
 {
-void DeltaMergeStore::setUpBackgroundTask(const DMContextPtr & dm_context)
+
+// A callback class for scanning the DMFiles on local filesystem
+class LocalDMFileGcScanner final
 {
-    // Callbacks for cleaning outdated DTFiles. Note that there is a chance
-    // that callbacks is called after the `DeltaMergeStore` dropped, we must
-    // make the callbacks safe.
-    ExternalPageCallbacks callbacks;
-    callbacks.ns_id = storage_pool->getNamespaceId();
-    callbacks.scanner = [path_pool_weak_ref = std::weak_ptr<StoragePathPool>(path_pool), file_provider = global_context.getFileProvider()]() {
+private:
+    // !!! Warning !!!
+    // Should only keep a weak ref of storage path pool since
+    // this callback instance may still valid inside the PageStorage
+    // even after the DeltaMerge storage is shutdown or released.
+    std::weak_ptr<StoragePathPool> path_pool_weak_ref;
+    FileProviderPtr file_provider;
+
+public:
+    LocalDMFileGcScanner(std::weak_ptr<StoragePathPool> path_pool_, FileProviderPtr provider)
+        : path_pool_weak_ref(std::move(path_pool_))
+        , file_provider(std::move(provider))
+    {}
+
+    ExternalPageCallbacks::PathAndIdsVec operator()()
+    {
         ExternalPageCallbacks::PathAndIdsVec path_and_ids_vec;
 
-        // If the StoragePathPool is invalid, meaning we call `scanner` after dropping the table,
+        // If the StoragePathPool is invalid or shutdown flag is set,
+        // meaning we call `scanner` after shutdowning or dropping the table,
         // simply return an empty list is OK.
         auto path_pool = path_pool_weak_ref.lock();
-        if (!path_pool)
+        if (!path_pool || path_pool->isShutdown())
             return path_and_ids_vec;
 
         // Return the DTFiles on disks.
@@ -66,14 +82,35 @@ void DeltaMergeStore::setUpBackgroundTask(const DMContextPtr & dm_context)
             path_and_ids_vec.emplace_back(root_path, std::move(file_ids_in_current_path));
         }
         return path_and_ids_vec;
-    };
-    callbacks.remover = [path_pool_weak_ref = std::weak_ptr<StoragePathPool>(path_pool), //
-                         file_provider = global_context.getFileProvider(),
-                         logger = log](const ExternalPageCallbacks::PathAndIdsVec & path_and_ids_vec, const std::set<PageId> & valid_ids) {
-        // If the StoragePathPool is invalid, meaning we call `remover` after dropping the table,
-        // simply skip is OK.
+    }
+};
+
+// A callback class for removing the DMFiles on local filesystem
+class LocalDMFileGcRemover final
+{
+private:
+    // !!! Warning !!!
+    // Should only keep a weak ref of storage path pool since
+    // this callback instance may still valid inside the PageStorage
+    // even after the DeltaMerge storage is shutdown or released.
+    std::weak_ptr<StoragePathPool> path_pool_weak_ref;
+    FileProviderPtr file_provider;
+    LoggerPtr logger;
+
+public:
+    LocalDMFileGcRemover(std::weak_ptr<StoragePathPool> path_pool_, FileProviderPtr provider, LoggerPtr log)
+        : path_pool_weak_ref(std::move(path_pool_))
+        , file_provider(std::move(provider))
+        , logger(std::move(log))
+    {}
+
+    void operator()(const ExternalPageCallbacks::PathAndIdsVec & path_and_ids_vec, const std::set<PageId> & valid_ids)
+    {
+        // If the StoragePathPool is invalid or shutdown flag is set,
+        // meaning we call `remover` after shutdowning or dropping the table,
+        // we must skip because the `valid_ids` is not reliable!
         auto path_pool = path_pool_weak_ref.lock();
-        if (!path_pool)
+        if (!path_pool || path_pool->isShutdown())
             return;
 
         SYNC_FOR("before_DeltaMergeStore::callbacks_remover_remove");
@@ -102,6 +139,7 @@ void DeltaMergeStore::setUpBackgroundTask(const DMContextPtr & dm_context)
                     LOG_INFO(logger,
                              "GC try remove useless DM file, but file not found and may have been removed, dmfile={}",
                              DMFile::getPathByStatus(path, id, DMFile::Status::READABLE));
+                    continue; // next file
                 }
                 else if (dmfile->canGC())
                 {
@@ -126,19 +164,30 @@ void DeltaMergeStore::setUpBackgroundTask(const DMContextPtr & dm_context)
                         LOG_INFO(logger, "GC removed useless DM file, dmfile={}", dmfile->path());
                     else
                         LOG_INFO(logger, "GC try remove useless DM file, but error happen, dmfile={} err_msg={}", dmfile->path(), err_msg);
+                    continue; // next file
                 }
             }
         }
-    };
+    }
+};
+
+void DeltaMergeStore::setUpBackgroundTask(const DMContextPtr & dm_context)
+{
+    // Callbacks for cleaning outdated DTFiles. Note that there is a chance
+    // that callbacks is called after the `DeltaMergeStore` shutdown or dropped,
+    // we must make the callbacks safe.
+    ExternalPageCallbacks callbacks;
+    callbacks.ns_id = storage_pool->getNamespaceId();
+    callbacks.scanner = LocalDMFileGcScanner(std::weak_ptr<StoragePathPool>(path_pool), global_context.getFileProvider());
+    callbacks.remover = LocalDMFileGcRemover(std::weak_ptr<StoragePathPool>(path_pool), global_context.getFileProvider(), log);
     // remember to unregister it when shutdown
-    storage_pool->dataRegisterExternalPagesCallbacks(callbacks);
-    storage_pool->enableGC();
+    storage_pool->startup(std::move(callbacks));
 
     background_task_handle = background_pool.addTask([this] { return handleBackgroundTask(false); });
 
     blockable_background_pool_handle = blockable_background_pool.addTask([this] { return handleBackgroundTask(true); });
 
-    // Do place delta index.
+    // Generate place delta index tasks
     for (auto & [end, segment] : segments)
     {
         (void)end;

--- a/dbms/src/Storages/DeltaMerge/StoragePool.cpp
+++ b/dbms/src/Storages/DeltaMerge/StoragePool.cpp
@@ -112,11 +112,7 @@ GlobalStoragePool::GlobalStoragePool(const PathPool & path_pool, Context & globa
 
 GlobalStoragePool::~GlobalStoragePool()
 {
-    if (gc_handle)
-    {
-        global_context.getBackgroundPool().removeTask(gc_handle);
-        gc_handle = nullptr;
-    }
+    shutdown();
 }
 
 void GlobalStoragePool::restore()
@@ -130,6 +126,15 @@ void GlobalStoragePool::restore()
             return this->gc(global_context.getSettingsRef());
         },
         false);
+}
+
+void GlobalStoragePool::shutdown()
+{
+    if (gc_handle)
+    {
+        global_context.getBackgroundPool().removeTask(gc_handle);
+        gc_handle = {};
+    }
 }
 
 FileUsageStatistics GlobalStoragePool::getLogFileUsage() const
@@ -535,30 +540,31 @@ StoragePool::~StoragePool()
     shutdown();
 }
 
-void StoragePool::enableGC()
-{
-    // The data in V3 will be GCed by `GlobalStoragePool::gc`, only register gc task under only v2/mix mode
-    if (run_mode == PageStorageRunMode::ONLY_V2 || run_mode == PageStorageRunMode::MIX_MODE)
-    {
-        gc_handle = global_context.getBackgroundPool().addTask([this] { return this->gc(global_context.getSettingsRef()); });
-    }
-}
-
-void StoragePool::dataRegisterExternalPagesCallbacks(const ExternalPageCallbacks & callbacks)
+void StoragePool::startup(ExternalPageCallbacks && callbacks)
 {
     switch (run_mode)
     {
     case PageStorageRunMode::ONLY_V2:
     {
+        // For V2, we need a per physical table gc handle to perform the gc of its PageStorage instances.
         data_storage_v2->registerExternalPagesCallbacks(callbacks);
+        gc_handle = global_context.getBackgroundPool().addTask([this] { return this->gc(global_context.getSettingsRef()); });
         break;
     }
     case PageStorageRunMode::ONLY_V3:
+    {
+        // For V3, the GC is handled by `GlobalStoragePool::gc`, just register callbacks is OK.
+        data_storage_v3->registerExternalPagesCallbacks(callbacks);
+        break;
+    }
     case PageStorageRunMode::MIX_MODE:
     {
-        // We have transformed all pages from V2 to V3 in `restore`, so
-        // only need to register callbacks for V3.
+        // For V3, the GC is handled by `GlobalStoragePool::gc`.
+        // Since we have transformed all external pages from V2 to V3 in `StoragePool::restore`,
+        // just register callbacks to V3 is OK
         data_storage_v3->registerExternalPagesCallbacks(callbacks);
+        // we still need a gc_handle to reclaim the V2 disk space.
+        gc_handle = global_context.getBackgroundPool().addTask([this] { return this->gc(global_context.getSettingsRef()); });
         break;
     }
     default:
@@ -566,19 +572,36 @@ void StoragePool::dataRegisterExternalPagesCallbacks(const ExternalPageCallbacks
     }
 }
 
-void StoragePool::dataUnregisterExternalPagesCallbacks(NamespaceId ns_id)
+void StoragePool::shutdown()
 {
+    // Note: Should reset the gc_handle before unregistering the pages callbacks
+    if (gc_handle)
+    {
+        global_context.getBackgroundPool().removeTask(gc_handle);
+        gc_handle = nullptr;
+    }
+
     switch (run_mode)
     {
     case PageStorageRunMode::ONLY_V2:
     {
+        meta_storage_v2->shutdown();
+        log_storage_v2->shutdown();
+        data_storage_v2->shutdown();
         data_storage_v2->unregisterExternalPagesCallbacks(ns_id);
         break;
     }
     case PageStorageRunMode::ONLY_V3:
+    {
+        data_storage_v3->unregisterExternalPagesCallbacks(ns_id);
+        break;
+    }
     case PageStorageRunMode::MIX_MODE:
     {
-        // We have transformed all pages from V2 to V3 in `restore`, so
+        meta_storage_v2->shutdown();
+        log_storage_v2->shutdown();
+        data_storage_v2->shutdown();
+        // We have transformed all external pages from V2 to V3 in `restore`, so
         // only need to unregister callbacks for V3.
         data_storage_v3->unregisterExternalPagesCallbacks(ns_id);
         break;
@@ -587,7 +610,6 @@ void StoragePool::dataUnregisterExternalPagesCallbacks(NamespaceId ns_id)
         throw Exception(fmt::format("Unknown PageStorageRunMode {}", static_cast<UInt8>(run_mode)), ErrorCodes::LOGICAL_ERROR);
     }
 }
-
 
 bool StoragePool::doV2Gc(const Settings & settings)
 {
@@ -627,21 +649,6 @@ bool StoragePool::gc(const Settings & settings, const Seconds & try_gc_period)
 
     // Only do the v2 GC
     return doV2Gc(settings);
-}
-
-void StoragePool::shutdown()
-{
-    if (gc_handle)
-    {
-        global_context.getBackgroundPool().removeTask(gc_handle);
-        gc_handle = nullptr;
-    }
-    if (run_mode != PageStorageRunMode::ONLY_V3)
-    {
-        meta_storage_v2->shutdown();
-        log_storage_v2->shutdown();
-        data_storage_v2->shutdown();
-    }
 }
 
 void StoragePool::drop()

--- a/dbms/src/Storages/DeltaMerge/StoragePool.h
+++ b/dbms/src/Storages/DeltaMerge/StoragePool.h
@@ -14,7 +14,6 @@
 
 #pragma once
 
-#include <Poco/Logger.h>
 #include <Storages/BackgroundProcessingPool.h>
 #include <Storages/Page/FileUsage.h>
 #include <Storages/Page/PageStorage.h>
@@ -36,7 +35,7 @@ namespace DM
 class StoragePool;
 using StoragePoolPtr = std::shared_ptr<StoragePool>;
 
-static const std::chrono::seconds DELTA_MERGE_GC_PERIOD(60);
+static constexpr std::chrono::seconds DELTA_MERGE_GC_PERIOD(60);
 
 class GlobalStoragePool : private boost::noncopyable
 {
@@ -50,6 +49,8 @@ public:
     ~GlobalStoragePool();
 
     void restore();
+
+    void shutdown();
 
     friend class StoragePool;
     friend class ::DB::AsynchronousMetrics;
@@ -90,7 +91,7 @@ public:
 
     NamespaceId getNamespaceId() const { return ns_id; }
 
-    PageStorageRunMode getPageStorageRunMode()
+    PageStorageRunMode getPageStorageRunMode() const
     {
         return run_mode;
     }
@@ -141,15 +142,14 @@ public:
     PageReader newMetaReader(ReadLimiterPtr read_limiter, bool snapshot_read, const String & tracing_id);
     PageReader newMetaReader(ReadLimiterPtr read_limiter, PageStorage::SnapshotPtr & snapshot);
 
-    void enableGC();
+    // Register the clean up DMFiles callbacks to PageStorage.
+    // The callbacks will be unregister when `shutdown` is called.
+    void startup(ExternalPageCallbacks && callbacks);
 
-    void dataRegisterExternalPagesCallbacks(const ExternalPageCallbacks & callbacks);
-
-    void dataUnregisterExternalPagesCallbacks(NamespaceId ns_id);
+    // Shutdown the gc handle and DMFile callbacks
+    void shutdown();
 
     bool gc(const Settings & settings, const Seconds & try_gc_period = DELTA_MERGE_GC_PERIOD);
-
-    void shutdown();
 
     // Caller must cancel gc tasks before drop
     void drop();

--- a/dbms/src/Storages/DeltaMerge/tests/gtest_dm_delta_merge_store.cpp
+++ b/dbms/src/Storages/DeltaMerge/tests/gtest_dm_delta_merge_store.cpp
@@ -97,6 +97,59 @@ try
 }
 CATCH
 
+TEST_F(DeltaMergeStoreTest, ShutdownInMiddleDTFileGC)
+try
+{
+    // create table
+    ASSERT_NE(store, nullptr);
+
+    auto global_page_storage = TiFlashTestEnv::getGlobalContext().getGlobalStoragePool();
+
+    // Start a PageStorage gc and suspend it before clean external page
+    auto sp_gc = SyncPointCtl::enableInScope("before_PageStorageImpl::cleanExternalPage_execute_callbacks");
+    auto th_gc = std::async([&]() {
+        if (global_page_storage)
+            global_page_storage->gc();
+    });
+    sp_gc.waitAndPause();
+
+    {
+        // check column structure of store
+        const auto & cols = store->getTableColumns();
+        // version & tag column added
+        ASSERT_EQ(cols.size(), 3);
+    }
+
+    auto provider = db_context->getFileProvider();
+    const auto paths = getAllStorePaths();
+    auto get_num_stable_files = [&]() -> size_t {
+        size_t total_num_dtfiles = 0;
+        auto options = DMFile::ListOptions{.only_list_can_gc = true, .clean_up = false};
+        for (const auto & root_path : paths)
+        {
+            auto file_ids = DMFile::listAllInPath(provider, root_path, options);
+            total_num_dtfiles += file_ids.size();
+        }
+        return total_num_dtfiles;
+    };
+
+    const size_t num_before_shutdown = get_num_stable_files();
+    ASSERT_GT(num_before_shutdown, 0);
+
+    // shutdown the table in the middle of page storage gc, but not dropped
+    store->shutdown();
+
+    const size_t num_after_shutdown = get_num_stable_files();
+    ASSERT_EQ(num_before_shutdown, num_after_shutdown);
+
+    sp_gc.next(); // continue the page storage gc
+    th_gc.get();
+
+    const size_t num_after_gc = get_num_stable_files();
+    ASSERT_EQ(num_before_shutdown, num_after_gc);
+}
+CATCH
+
 TEST_F(DeltaMergeStoreTest, DroppedInMiddleDTFileGC)
 try
 {
@@ -126,6 +179,8 @@ try
 
     sp_gc.next(); // continue the page storage gc
     th_gc.get();
+
+    // Mark: ensure this test won't run into crash by nullptr
 }
 CATCH
 

--- a/dbms/src/Storages/DeltaMerge/tests/gtest_dm_delta_merge_store_test_basic.h
+++ b/dbms/src/Storages/DeltaMerge/tests/gtest_dm_delta_merge_store_test_basic.h
@@ -77,6 +77,12 @@ public:
         return s;
     }
 
+    Strings getAllStorePaths() const
+    {
+        auto path_delegate = store->path_pool->getStableDiskDelegator();
+        return path_delegate.listPaths();
+    }
+
 protected:
     DeltaMergeStorePtr store;
 };

--- a/dbms/src/Storages/Page/V2/PageStorage.cpp
+++ b/dbms/src/Storages/Page/V2/PageStorage.cpp
@@ -621,6 +621,7 @@ size_t PageStorage::getNumberOfPages()
     }
 }
 
+// For debugging purpose
 std::set<PageId> PageStorage::getAliveExternalPageIds(NamespaceId /*ns_id*/)
 {
     const auto & concrete_snap = getConcreteSnapshot();

--- a/dbms/src/Storages/Page/V2/PageStorage.h
+++ b/dbms/src/Storages/Page/V2/PageStorage.h
@@ -92,7 +92,7 @@ public:
                 const FileProviderPtr & file_provider_,
                 BackgroundProcessingPool & ver_compact_pool_,
                 bool no_more_insert_ = false);
-    ~PageStorage() override { shutdown(); }
+    ~PageStorage() override { shutdown(); } // NOLINT(clang-analyzer-optin.cplusplus.VirtualCall)
 
     void restore() override;
 

--- a/dbms/src/Storages/Page/V3/PageDirectory.h
+++ b/dbms/src/Storages/Page/V3/PageDirectory.h
@@ -339,7 +339,7 @@ public:
 
     // Get the external id that is not deleted or being ref by another id by
     // `ns_id`.
-    std::set<PageId> getAliveExternalIds(NamespaceId ns_id) const
+    std::optional<std::set<PageId>> getAliveExternalIds(NamespaceId ns_id) const
     {
         return external_ids_by_ns.getAliveIds(ns_id);
     }

--- a/dbms/src/Storages/Page/V3/PageDirectory/ExternalIdsByNamespace.cpp
+++ b/dbms/src/Storages/Page/V3/PageDirectory/ExternalIdsByNamespace.cpp
@@ -16,6 +16,7 @@
 #include <Storages/Page/V3/PageDirectory/ExternalIdsByNamespace.h>
 
 #include <mutex>
+#include <optional>
 
 namespace DB::PS::V3
 {
@@ -34,17 +35,17 @@ void ExternalIdsByNamespace::addExternalId(const std::shared_ptr<PageIdV3Interna
     addExternalIdUnlock(external_id);
 }
 
-std::set<PageId> ExternalIdsByNamespace::getAliveIds(NamespaceId ns_id) const
+std::optional<std::set<PageId>> ExternalIdsByNamespace::getAliveIds(NamespaceId ns_id) const
 {
     // Now we assume a lock among all NamespaceIds is good enough.
     std::unique_lock map_guard(mu);
 
-    std::set<PageId> valid_external_ids;
     auto ns_iter = ids_by_ns.find(ns_id);
     if (ns_iter == ids_by_ns.end())
-        return valid_external_ids;
+        return std::nullopt;
 
     // Only scan the given `ns_id`
+    std::set<PageId> valid_external_ids;
     auto & external_ids = ns_iter->second;
     for (auto iter = external_ids.begin(); iter != external_ids.end(); /*empty*/)
     {
@@ -61,11 +62,8 @@ std::set<PageId> ExternalIdsByNamespace::getAliveIds(NamespaceId ns_id) const
             ++iter;
         }
     }
-    // No valid external pages in this `ns_id`
-    if (valid_external_ids.empty())
-    {
-        ids_by_ns.erase(ns_id);
-    }
+    // The `external_ids` maybe an empty list now, leave it to be
+    // cleaned by unregister
     return valid_external_ids;
 }
 

--- a/dbms/src/Storages/Page/V3/PageDirectory/ExternalIdsByNamespace.h
+++ b/dbms/src/Storages/Page/V3/PageDirectory/ExternalIdsByNamespace.h
@@ -40,7 +40,8 @@ public:
 
     // Get all alive external ids of given `ns_id`
     // Will also cleanup the invalid external ids.
-    std::set<PageId> getAliveIds(NamespaceId ns_id) const;
+    // If the ns_id is invalid, std::nullopt will be returned.
+    std::optional<std::set<PageId>> getAliveIds(NamespaceId ns_id) const;
 
     // After table dropped, the `getAliveIds` with specified
     // `ns_id` will not be cleaned. We need this method to

--- a/dbms/src/Storages/Page/V3/PageStorageImpl.cpp
+++ b/dbms/src/Storages/Page/V3/PageStorageImpl.cpp
@@ -115,9 +115,13 @@ size_t PageStorageImpl::getNumberOfPages()
     return page_directory->numPages();
 }
 
+// For debugging purpose
 std::set<PageId> PageStorageImpl::getAliveExternalPageIds(NamespaceId ns_id)
 {
-    return page_directory->getAliveExternalIds(ns_id);
+    // Keep backward compatibility of this functions with v2
+    if (auto ids = page_directory->getAliveExternalIds(ns_id); ids)
+        return *ids;
+    return {};
 }
 
 void PageStorageImpl::writeImpl(DB::WriteBatch && write_batch, const WriteLimiterPtr & write_limiter)
@@ -369,8 +373,11 @@ void PageStorageImpl::cleanExternalPage(Stopwatch & gc_watch, GCTimeStatistics &
         statistics.external_page_scan_ns += external_watch.elapsedFromLastTime();
         auto alive_external_ids = page_directory->getAliveExternalIds(ns_callbacks->ns_id);
         statistics.external_page_get_alive_ns += external_watch.elapsedFromLastTime();
-        // remove the external pages that is not alive now.
-        ns_callbacks->remover(pending_external_pages, alive_external_ids);
+        if (alive_external_ids)
+        {
+            // remove the external pages that is not alive now.
+            ns_callbacks->remover(pending_external_pages, *alive_external_ids);
+        } // else the ns_id is invalid, just skip
         statistics.external_page_remove_ns += external_watch.elapsedFromLastTime();
 
         // move to next namespace callbacks

--- a/dbms/src/Storages/Page/V3/tests/gtest_page_directory.cpp
+++ b/dbms/src/Storages/Page/V3/tests/gtest_page_directory.cpp
@@ -81,9 +81,10 @@ TEST(ExternalIdsByNamespaceTest, Simple)
     {
         // holder keep "50" alive
         auto ids = external_ids_by_ns.getAliveIds(ns_id);
-        LOG_DEBUG(&Poco::Logger::root(), "{} end first, size={}", who.load(), ids.size());
-        ASSERT_EQ(ids.size(), 1);
-        ASSERT_EQ(*ids.begin(), 50);
+        ASSERT_TRUE(ids.has_value());
+        LOG_DEBUG(Logger::get(), "{} end first, size={}", who.load(), ids->size());
+        ASSERT_EQ(ids->size(), 1);
+        ASSERT_EQ(*ids->begin(), 50);
         ASSERT_TRUE(external_ids_by_ns.existNamespace(ns_id));
     }
 
@@ -92,7 +93,7 @@ TEST(ExternalIdsByNamespaceTest, Simple)
         // "50" is erased though the holder is not released.
         external_ids_by_ns.unregisterNamespace(ns_id);
         auto ids = external_ids_by_ns.getAliveIds(ns_id);
-        ASSERT_EQ(ids.size(), 0);
+        ASSERT_FALSE(ids.has_value()); // nullopt
         ASSERT_FALSE(external_ids_by_ns.existNamespace(ns_id));
     }
 }
@@ -547,8 +548,9 @@ TEST_F(PageDirectoryTest, IdempotentNewExtPageAfterAllCleaned)
         edit.putExternal(buildV3Id(TEST_NAMESPACE_ID, 10));
         dir->apply(std::move(edit));
         auto alive_ids = dir->getAliveExternalIds(TEST_NAMESPACE_ID);
-        EXPECT_EQ(alive_ids.size(), 1);
-        EXPECT_GT(alive_ids.count(10), 0);
+        ASSERT_TRUE(alive_ids.has_value());
+        EXPECT_EQ(alive_ids->size(), 1);
+        EXPECT_GT(alive_ids->count(10), 0);
     }
 
     {
@@ -556,8 +558,9 @@ TEST_F(PageDirectoryTest, IdempotentNewExtPageAfterAllCleaned)
         edit.putExternal(buildV3Id(TEST_NAMESPACE_ID, 10)); // should be idempotent
         dir->apply(std::move(edit));
         auto alive_ids = dir->getAliveExternalIds(TEST_NAMESPACE_ID);
-        EXPECT_EQ(alive_ids.size(), 1);
-        EXPECT_GT(alive_ids.count(10), 0);
+        ASSERT_TRUE(alive_ids.has_value());
+        EXPECT_EQ(alive_ids->size(), 1);
+        EXPECT_GT(alive_ids->count(10), 0);
     }
 
     {
@@ -566,8 +569,9 @@ TEST_F(PageDirectoryTest, IdempotentNewExtPageAfterAllCleaned)
         dir->apply(std::move(edit));
         dir->gcInMemEntries(); // clean in memory
         auto alive_ids = dir->getAliveExternalIds(TEST_NAMESPACE_ID);
-        EXPECT_EQ(alive_ids.size(), 0);
-        EXPECT_EQ(alive_ids.count(10), 0); // removed
+        ASSERT_TRUE(alive_ids.has_value());
+        EXPECT_EQ(alive_ids->size(), 0);
+        EXPECT_EQ(alive_ids->count(10), 0); // removed
     }
 
     {
@@ -576,8 +580,9 @@ TEST_F(PageDirectoryTest, IdempotentNewExtPageAfterAllCleaned)
         edit.putExternal(buildV3Id(TEST_NAMESPACE_ID, 10));
         dir->apply(std::move(edit));
         auto alive_ids = dir->getAliveExternalIds(TEST_NAMESPACE_ID);
-        EXPECT_EQ(alive_ids.size(), 1);
-        EXPECT_GT(alive_ids.count(10), 0);
+        ASSERT_TRUE(alive_ids.has_value());
+        EXPECT_EQ(alive_ids->size(), 1);
+        EXPECT_GT(alive_ids->count(10), 0);
     }
 }
 
@@ -915,8 +920,9 @@ try
         }
         auto snap = dir->createSnapshot();
         auto alive_ids = dir->getAliveExternalIds(TEST_NAMESPACE_ID);
-        EXPECT_EQ(alive_ids.size(), 1);
-        EXPECT_GT(alive_ids.count(50), 0);
+        ASSERT_TRUE(alive_ids.has_value());
+        EXPECT_EQ(alive_ids->size(), 1);
+        EXPECT_GT(alive_ids->count(50), 0);
     }
 }
 CATCH
@@ -1732,8 +1738,9 @@ try
         auto outdated_entries = dir->gcInMemEntries();
         EXPECT_TRUE(outdated_entries.empty());
         auto alive_ex_id = dir->getAliveExternalIds(TEST_NAMESPACE_ID);
-        ASSERT_EQ(alive_ex_id.size(), 1);
-        ASSERT_EQ(*alive_ex_id.begin(), 10);
+        ASSERT_TRUE(alive_ex_id.has_value());
+        ASSERT_EQ(alive_ex_id->size(), 1);
+        ASSERT_EQ(*alive_ex_id->begin(), 10);
     }
 
     // del 11->ext
@@ -1747,7 +1754,8 @@ try
         auto outdated_entries = dir->gcInMemEntries();
         EXPECT_EQ(0, outdated_entries.size());
         auto alive_ex_id = dir->getAliveExternalIds(TEST_NAMESPACE_ID);
-        ASSERT_EQ(alive_ex_id.size(), 0);
+        ASSERT_TRUE(alive_ex_id.has_value());
+        ASSERT_EQ(alive_ex_id->size(), 0);
     }
 }
 CATCH
@@ -1889,10 +1897,11 @@ try
         EXPECT_ANY_THROW(getEntry(restored_dir, 2, temp_snap));
         EXPECT_ANY_THROW(getEntry(restored_dir, 3, temp_snap));
         auto alive_ex = restored_dir->getAliveExternalIds(TEST_NAMESPACE_ID);
-        EXPECT_EQ(alive_ex.size(), 3);
-        EXPECT_GT(alive_ex.count(10), 0);
-        EXPECT_GT(alive_ex.count(20), 0);
-        EXPECT_GT(alive_ex.count(30), 0);
+        ASSERT_TRUE(alive_ex.has_value());
+        EXPECT_EQ(alive_ex->size(), 3);
+        EXPECT_GT(alive_ex->count(10), 0);
+        EXPECT_GT(alive_ex->count(20), 0);
+        EXPECT_GT(alive_ex->count(30), 0);
         EXPECT_SAME_ENTRY(getEntry(restored_dir, 50, temp_snap), entry_50);
         EXPECT_SAME_ENTRY(getEntry(restored_dir, 60, temp_snap), entry_60);
     };
@@ -1927,15 +1936,16 @@ try
         EXPECT_ANY_THROW(getEntry(restored_dir, 2, temp_snap));
         EXPECT_ANY_THROW(getEntry(restored_dir, 3, temp_snap));
         auto alive_ex = restored_dir->getAliveExternalIds(TEST_NAMESPACE_ID);
-        EXPECT_EQ(alive_ex.size(), 2);
-        EXPECT_GT(alive_ex.count(10), 0);
+        ASSERT_TRUE(alive_ex.has_value());
+        EXPECT_EQ(alive_ex->size(), 2);
+        EXPECT_GT(alive_ex->count(10), 0);
         EXPECT_EQ(getNormalPageIdU64(restored_dir, 11, temp_snap), 10);
 
-        EXPECT_GT(alive_ex.count(20), 0);
+        EXPECT_GT(alive_ex->count(20), 0);
         EXPECT_EQ(getNormalPageIdU64(restored_dir, 21, temp_snap), 20);
         EXPECT_EQ(getNormalPageIdU64(restored_dir, 22, temp_snap), 20);
 
-        EXPECT_EQ(alive_ex.count(30), 0); // removed
+        EXPECT_EQ(alive_ex->count(30), 0); // removed
 
         EXPECT_ANY_THROW(getEntry(restored_dir, 50, temp_snap));
         EXPECT_SAME_ENTRY(getEntry(restored_dir, 51, temp_snap), entry_50);
@@ -1967,10 +1977,11 @@ try
         EXPECT_ANY_THROW(getEntry(restored_dir, 2, temp_snap));
         EXPECT_ANY_THROW(getEntry(restored_dir, 3, temp_snap));
         auto alive_ex = restored_dir->getAliveExternalIds(TEST_NAMESPACE_ID);
-        EXPECT_EQ(alive_ex.size(), 0);
-        EXPECT_EQ(alive_ex.count(10), 0); // removed
-        EXPECT_EQ(alive_ex.count(20), 0); // removed
-        EXPECT_EQ(alive_ex.count(30), 0); // removed
+        ASSERT_TRUE(alive_ex.has_value());
+        EXPECT_EQ(alive_ex->size(), 0);
+        EXPECT_EQ(alive_ex->count(10), 0); // removed
+        EXPECT_EQ(alive_ex->count(20), 0); // removed
+        EXPECT_EQ(alive_ex->count(30), 0); // removed
 
         EXPECT_ANY_THROW(getEntry(restored_dir, 50, temp_snap));
         EXPECT_SAME_ENTRY(getEntry(restored_dir, 51, temp_snap), entry_50);
@@ -1999,10 +2010,11 @@ try
         EXPECT_ANY_THROW(getEntry(restored_dir, 2, temp_snap));
         EXPECT_ANY_THROW(getEntry(restored_dir, 3, temp_snap));
         auto alive_ex = restored_dir->getAliveExternalIds(TEST_NAMESPACE_ID);
-        EXPECT_EQ(alive_ex.size(), 0);
-        EXPECT_EQ(alive_ex.count(10), 0); // removed
-        EXPECT_EQ(alive_ex.count(20), 0); // removed
-        EXPECT_EQ(alive_ex.count(30), 0); // removed
+        ASSERT_TRUE(alive_ex.has_value());
+        EXPECT_EQ(alive_ex->size(), 0);
+        EXPECT_EQ(alive_ex->count(10), 0); // removed
+        EXPECT_EQ(alive_ex->count(20), 0); // removed
+        EXPECT_EQ(alive_ex->count(30), 0); // removed
 
         EXPECT_ANY_THROW(getEntry(restored_dir, 50, temp_snap));
         EXPECT_ANY_THROW(getEntry(restored_dir, 51, temp_snap));

--- a/dbms/src/Storages/Page/V3/tests/gtest_page_storage.cpp
+++ b/dbms/src/Storages/Page/V3/tests/gtest_page_storage.cpp
@@ -1265,9 +1265,22 @@ CATCH
 TEST_F(PageStorageTest, ConcurrencyAddExtCallbacks)
 try
 {
+    NamespaceId ns_id1 = TEST_NAMESPACE_ID;
+    NamespaceId ns_id2 = TEST_NAMESPACE_ID + 1;
+    {
+        WriteBatch wb(ns_id1);
+        wb.putExternal(20, 0);
+        page_storage->write(std::move(wb));
+    }
+    {
+        WriteBatch wb(ns_id2);
+        wb.putExternal(20, 0);
+        page_storage->write(std::move(wb));
+    }
+
     auto ptr = std::make_shared<Int32>(100); // mock the `StorageDeltaMerge`
     ExternalPageCallbacks callbacks;
-    callbacks.ns_id = TEST_NAMESPACE_ID;
+    callbacks.ns_id = ns_id1;
     callbacks.scanner = [ptr_weak_ref = std::weak_ptr<Int32>(ptr)]() -> ExternalPageCallbacks::PathAndIdsVec {
         auto ptr = ptr_weak_ref.lock();
         if (!ptr)
@@ -1295,7 +1308,7 @@ try
     // mock table created while gc is running
     {
         ExternalPageCallbacks new_callbacks;
-        new_callbacks.ns_id = TEST_NAMESPACE_ID + 1;
+        new_callbacks.ns_id = ns_id2;
         new_callbacks.scanner = [ptr_weak_ref = std::weak_ptr<Int32>(ptr)]() -> ExternalPageCallbacks::PathAndIdsVec {
             auto ptr = ptr_weak_ref.lock();
             if (!ptr)

--- a/dbms/src/Storages/PathPool.cpp
+++ b/dbms/src/Storages/PathPool.cpp
@@ -134,6 +134,7 @@ StoragePathPool::StoragePathPool( //
     : database(std::move(database_))
     , table(std::move(table_))
     , path_need_database_name(path_need_database_name_)
+    , shutdown_called(false)
     , global_capacity(std::move(global_capacity_))
     , file_provider(std::move(file_provider_))
     , log(Logger::get("StoragePathPool"))
@@ -161,6 +162,7 @@ StoragePathPool::StoragePathPool(StoragePathPool && rhs) noexcept
     , table(std::move(rhs.table))
     , dt_file_path_map(std::move(rhs.dt_file_path_map))
     , path_need_database_name(rhs.path_need_database_name)
+    , shutdown_called(rhs.shutdown_called.load())
     , global_capacity(std::move(rhs.global_capacity))
     , file_provider(std::move(rhs.file_provider))
     , log(std::move(rhs.log))
@@ -176,6 +178,7 @@ StoragePathPool & StoragePathPool::operator=(StoragePathPool && rhs)
         database.swap(rhs.database);
         table.swap(rhs.table);
         path_need_database_name = rhs.path_need_database_name;
+        shutdown_called = rhs.shutdown_called.load();
         global_capacity.swap(rhs.global_capacity);
         file_provider.swap(rhs.file_provider);
         log.swap(rhs.log);

--- a/dbms/src/Storages/PathPool.h
+++ b/dbms/src/Storages/PathPool.h
@@ -411,7 +411,7 @@ private:
     PathPool::PageFilePathMap page_path_map;
 };
 
-/// A class to manage paths for the specified storage.
+/// A class to manage paths for a specified physical table.
 class StoragePathPool
 {
 public:
@@ -447,6 +447,10 @@ public:
     void rename(const String & new_database, const String & new_table);
 
     void drop(bool recursive, bool must_success = true);
+
+    void shutdown() { shutdown_called.store(true); }
+
+    bool isShutdown() const { return shutdown_called.load(); }
 
     DISALLOW_COPY(StoragePathPool);
 
@@ -495,6 +499,8 @@ private:
     DMFilePathMap dt_file_path_map;
 
     bool path_need_database_name = false;
+
+    std::atomic<bool> shutdown_called;
 
     PathCapacityMetricsPtr global_capacity;
 

--- a/dbms/src/Storages/StorageDeltaMerge.cpp
+++ b/dbms/src/Storages/StorageDeltaMerge.cpp
@@ -1555,6 +1555,9 @@ void StorageDeltaMerge::startup()
     tmt.getStorages().put(std::static_pointer_cast<StorageDeltaMerge>(shared_from_this()));
 }
 
+// Avoid calling virtual function `shutdown` in destructor,
+// we should use this function instead.
+// https://stackoverflow.com/a/12093250/4412495
 void StorageDeltaMerge::shutdownImpl()
 {
     bool v = false;

